subsys/usb/host/usbh_ch9.c

```diff
@@ -33,139 +45,188 @@ int usbh_req_setup(const struct device *dev,
 		.wLength = sys_cpu_to_le16(wLength),
 	};
 	struct uhc_transfer *xfer;
-	struct net_buf *buf;
 	uint8_t ep = usb_reqtype_is_to_device(&req) ? 0x00 : 0x80;
 	int ret;
 
-	xfer = uhc_xfer_alloc(dev, addr, ep, 0, 64, SETUP_REQ_TIMEOUT, NULL, NULL);
+	xfer = usbh_xfer_alloc(udev, ep, 0, 64, SETUP_REQ_TIMEOUT, (void *)ch9_req_cb);
 	if (!xfer) {
 		return -ENOMEM;
 	}
 
 	memcpy(xfer->setup_pkt, &req, sizeof(req));
 
-	if (wLength) {
-		buf = uhc_xfer_buf_alloc(dev, wLength);
-		if (!buf) {
-			ret = -ENOMEM;
-			goto buf_alloc_err;
-		}
-
-		if (usb_reqtype_is_to_device(&req) && data != NULL) {
-			net_buf_add_mem(buf, data, wLength);
-		}
-
-		ret = uhc_xfer_buf_add(dev, xfer, buf);
+	if (wLength && buf != NULL) {
```

Comment:
I would prefer to have assert for `(data != NULL && wLength) || (data == NULL && !wLength)` before the if condition and then just check `wLength` inside the if. I prefer invalid code to crash on NULL pointer somewhere (if asserts are disabled) than to have the code silence the error and propagate it further down.

The way this is currently implemented, it is possible for invalid code (i.e. code that calls this function with clearly wrong parameters) to send wLength > 0 *and* have the host to proceed directly to status stage. Please make the future debugging easier and just assert.

Reply:
I will add commit to address it.